### ISSUE TITLE
Discussion: Inheritance and Decoration

### ISSUE THREAD

**Issue by Iheartweb** [Droppe/Lu#117]
This is an exploration. Let me know your thoughts.

``` js

var Lu = window.Lu = {};

( function(){

 function klass(){};
 function inherits( parent, fn ){
   var child,
       proto;

   if( fn && fn.hasOwnProperty( 'constructor' ) ) {
     child = fn.constructor;
   } else if( fn && typeof fn === 'function' ){
     child = fn;
     fn = { constructor: child };
   } else {
     child = function(){
       return parent.call( this, parent, fn );
     }
   }

   _.extend( child, parent );

   klass.prototype = parent.prototype;
   child.prototype = proto = new klass();

   if( fn ){
     _.extend( child.prototype, fn );
   }

   proto.constructor = child;

   proto.decorate = function( decorator ){
       for (var key in decorator) {
           if (decorator.hasOwnProperty(key)) {
               proto[key] = decorator[key];
           }
       }
       return this;
   }

   //underscore underscore, Wha?
   child.base = parent;

   return child;
 }

 //Facade for everything above
 function extend( proto ){
   var child = inherits( this, proto );
   child.extend = extend;
   return child;
 }

 Lu.Class = function(){};
 Lu.Class.extend = extend;

}() );



var Foo = Lu.Class.extend( function(){

 var MAGIC_NUMBER = 1;

 return {
   constructor: function( prop ){
     var Foo = this;

     this.prop = prop;

     function privt(){
       return MAGIC_NUMBER;
     }
     Foo.privl = function() {
       return MAGIC_NUMBER += 1;
     }
   }
 }
}() );

Foo.prototype.proto = function(){
 return false;
};

var Bar = Foo.extend( function(){
 return {
   constructor: function( prop ){
     this.baz = prop;
     Bar.base.constructor.call( this, prop );
   },
   statc: function(){
     console.log( 'hello' );
   }
 }

}() );

Bar.prototype.proto = function(){
  //Bar.base
}


var TheRock = Bar.extend( function(){
 this.awesome = true;
 function privatelyDoSomthing(){
   return 'it\'s all good.';
 }
} );

TheRock.prototype.eyebrowMove = function(){
 return 'Wha';
};

var instanceFoo = new Foo(),
 instanceBar = new Bar( 'heya' ),
 Kiro = new TheRock();


function Hero( obj ){

  function somethingPrivate() {
  }

  return {
    foo: function() {
      console.log('foo');
    },
    bar: function() {
      this.foo();
    }
  }
}

Kiro.decorate( Hero );

console.log( instanceFoo, instanceFoo instanceof Foo );
console.log( instanceBar, instanceBar instanceof Bar );
console.log( Kiro, Kiro instanceof TheRock );
```

**Comment by Jimgskoop** [comment]
Sorry, I'm coming into the middle of this thread. What's the rationale  
for moving to the decorator pattern for extension/inheritance at this  
point?

Thanks,

-Jim G

On May 4, 2012, at 2:39 PM, Robert Martone <reply@reply.github.com

**Comment by krisk** [comment]
Create question @Jimgskoop.

### A little history

This came about when we were looking at how some Lu controls had been implemented.  For example, consider the Tip control and how it is constructed:

``` html
<div data-lu="Tip" title="This is my tip" />
```

In its implementation, the Tip control essentially creates a Container control, in which it injects the text defined in the `title` attribute.  @Iheartweb and I initially thought that  instead of creating a Container, the tip itself should be a Container.  In other words, it should inherit from Container, as per the old adage, a tip IS a container.

After giving it more thought, we realized we were wrong.  A tip itself is merely a textual artifact, usually defined by an attribute (i.e, title) of an HTML node.  From this, we deduced that a tip should not be an instantiable control with a target DOM.  Basically, the HTML author should **not** have to introduce a tip element when designing a page.

Why? Because, aside from having to maintain additional DOM, something like an `<a>` tag isn't a tip, but rather may have a tip, hence the different adage: something HAS a tip.

As far as Lu is concerned, how do we present this HAS-A relationship?  Well, with decorators.  

### Decorators

Consider the following `<a>` tag:

``` html
<a data-lu="Button" title="This is my tip" />
```

With a decorator pattern, we merely construct the Lu button, and decorate it with a Tip (yes, there perhaps need to be more configuration points defined, such as orientation, appearance delay, etc..).  

Applying the decoration pattern to Lu control construction reduces the weight of the Lu framework.  Without the decoration pattern, we may end up fattening exiting, or creating new, controls, simply to add behavior.  Wouldn't it be simpler, cleaner, and (arguably) better to merely extend a control's interface with a separate object?  

### Inheritance + Decoration

Given the above (and from a Lu meeting 2 weeks ago), we wanted to come up with an inheritance model for Lu which would also facilitate decoration.  

In other words, we wanted to have something like the following HTML

``` html
<a data-lu="Button" title="This is my tip" />
```

be translated into this JavaScript

``` javascript
var Button = Lu.Class.extend( function () { ... }); // Define the Button control
var Tip = { ... };  // Define the tip object

var btn = new Button(); // Create a new Button
btn.decorate(Tip);      // Decorate the button instance with a tip
```

(or a similar effect.  Note: the interface has not been concretized just yet)

I hope this is clearer!

**Comment by Iheartweb** [comment]
@Jimgskoop It's not inheritance vs decoration. We are hoping for inheritance + decoration.

Decoration allows us to add functionality to already instantiated controls. This is much cheaper than creating a class that extend an object and as @krisk points out it makes a bit more sense in some situations.

inheritance means 'is a'
decoration means 'has a'

**Comment by Iheartweb** [comment]
Gents,

I've done some toying around with the various class libs and here's the gist.

The backbone-esc style (as in the above) does not easily support the inheritance of privileged methods. Calling `Foo.extend()` does not instantiate Foo and therefore its constructor is never called. So if you had:

``` JS
var Foo = Class.extend( function(){
  return {
    constructor: function(){
      this.theforce = function(){
        return true;
      }
    }
  };
}() );

var Bar = Foo.extend( /*object*/ ); 
```

Bar would not inherit `theforce`. It would only inherit methods that are on Foo's prototype.

On the other hand we do have access to the constructor: `Foo.prototype.constructor` and we can instantiate it in Bar's constructor and then copy the methods to Bar. But, to make this clean is not an easy thing to do. 

PTClass gets around this by making `$super` avaliable as a param. Calling `$super` does the above: calls the parents constructor and then copies over methods.

klass auto instantiates the superclass and copies its methods automagially.

As you may not always need the superclass to be constructed and as you may need to adpat settings, I prefer the PTclass route.

Here's were that leaves us:
1. We can refactor PTClass to add support for decoration (addMethods sort of does what we need). While in there we should use underscore for extension and utility. We also need to verify that ptclass set up the prototype chain correctly.
2. Or, refactor the above so that it mimics PTClass' $super.

Regardless of what we go with, we will still have to refactor Lu primitives to use more prototypical inheritance.

Thoughts?

**Comment by krisk** [comment]
All right gentlemen, how about something like the following:  https://gist.github.com/2631660

**Comment by krisk** [comment]
Lot's to talk about ....  completely re-writing inheritance... _ouch_.

Let's meet sometime tomorrow, whoever is available.

**Comment by Jimgskoop** [comment]
I nominate Kiro to write all documentation from now on!  ;)

**Comment by Jimgskoop** [comment]
Question:  Is now the right time to deal with overhauling Lu's guts?  Would it be more prudent to do this after we reach a usable v1.0 release?  Should our main goal right now be to get Lu usable on Davinci, and worry about enhancements later?

**Comment by Iheartweb** [comment]
This is a must before we allow people to iterate with Lu. And, saves us time as it greatly simplifies the creation of controls.  It is ambitious, but given "20%" we can handle it.

**Comment by krisk** [comment]
All right, yet another inheritance model:  https://gist.github.com/2646493

Let me know what you guys think.  Rob and I believe it covers pretty much all use cases.

**Comment by krisk** [comment]
And another one! https://gist.github.com/2667487

I believe we're going to go with this one.
